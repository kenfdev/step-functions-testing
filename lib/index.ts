import { Construct } from 'constructs';
// import * as sqs from 'aws-cdk-lib/aws-sqs';

export interface StepFunctionsTestingProps {
  // Define construct properties here
}

export class StepFunctionsTesting extends Construct {

  constructor(scope: Construct, id: string, props: StepFunctionsTestingProps = {}) {
    super(scope, id);

    // Define construct contents here

    // example resource
    // const queue = new sqs.Queue(this, 'StepFunctionsTestingQueue', {
    //   visibilityTimeout: cdk.Duration.seconds(300)
    // });
  }
}
