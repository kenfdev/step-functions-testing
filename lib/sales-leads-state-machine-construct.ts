import { Construct } from 'constructs';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { JsonPath, TaskInput } from 'aws-cdk-lib/aws-stepfunctions';
import * as ddb from 'aws-cdk-lib/aws-dynamodb';
import { Duration } from 'aws-cdk-lib';
import { DynamoAttributeValue } from 'aws-cdk-lib/aws-stepfunctions-tasks';

type StateMachineProps = {
  stateMachineName?: string;
  validationStateName?: string;
  checkIdentityStateName?: string;
  checkAddressStateName?: string;
  customValidationFailedStateName?: string;
  validationExceptionStateName?: string;
  detectSentimentStateName?: string;
  isPositiveSentimentStateName?: string;
  negativeSentimentDetectedStateName?: string;
  addToFollowUpStateName?: string;
  customerAddedToFollowupStateName?: string;
  retryInterval?: Duration;
};

export class SalesLeadsStateMachineConstruct extends Construct {
  constructor(scope: Construct, id: string, props: StateMachineProps = {}) {
    super(scope, id);

    const {
      validationStateName = 'Validation',
      checkIdentityStateName = 'Check Identity',
      checkAddressStateName = 'Check Address',
      customValidationFailedStateName = 'CustomValidationFailed',
      validationExceptionStateName = 'ValidationException',
      detectSentimentStateName = 'DetectSentiment',
      isPositiveSentimentStateName = 'Is Positive Sentiment?',
      negativeSentimentDetectedStateName = 'NegativeSentimentDetected',
      addToFollowUpStateName = 'Add to FollowUp',
      customerAddedToFollowupStateName = 'CustomerAddedToFollowup',
      stateMachineName = 'SalesLeadStateMachine',
      retryInterval = Duration.seconds(1),
    } = props;

    const checkIdentityState = new tasks.LambdaInvoke(
      this,
      checkIdentityStateName,
      {
        lambdaFunction: lambda.Function.fromFunctionArn(
          this,
          'CheckIdentityFunction',
          'arn:aws:lambda:us-west-2:123456789012:function:check-identity-function'
        ),
        inputPath: JsonPath.stringAt('$.data.identity'),
        payload: TaskInput.fromJsonPathAt('$'),
        resultSelector: {
          identity: JsonPath.stringToJson(JsonPath.stringAt('$.Payload.body')),
        },
        retryOnServiceExceptions: false,
      }
    ).addRetry({
      errors: [
        'Lambda.ServiceException',
        'Lambda.AWSLambdaException',
        'Lambda.SdkClientException',
        'CustomValidationError',
      ],
      backoffRate: 1,
      maxAttempts: 3,
      interval: retryInterval,
    });

    const checkAddressState = new tasks.LambdaInvoke(
      this,
      checkAddressStateName,
      {
        lambdaFunction: lambda.Function.fromFunctionArn(
          this,
          'CheckAddressFunction',
          'arn:aws:lambda:us-west-2:123456789012:function:check-address-function'
        ),
        inputPath: '$.data.address',
        payload: TaskInput.fromJsonPathAt('$'),
        resultSelector: {
          address: JsonPath.stringToJson(JsonPath.stringAt('$.Payload.body')),
        },
        retryOnServiceExceptions: false,
      }
    ).addRetry({
      errors: [
        'Lambda.ServiceException',
        'Lambda.AWSLambdaException',
        'Lambda.SdkClientException',
        'CustomValidationError',
      ],
      backoffRate: 1,
      maxAttempts: 3,
      interval: retryInterval,
    });

    const customValidationFailedState = new tasks.EventBridgePutEvents(
      this,
      customValidationFailedStateName,
      {
        entries: [
          {
            detail: TaskInput.fromObject({
              Message: 'Validation Failed',
            }),
            detailType: 'ValidationFailed',
            source: 'LocalTestingSource',
          },
        ],
      }
    );

    const validationExceptionState = new tasks.EventBridgePutEvents(
      this,
      validationExceptionStateName,
      {
        entries: [
          {
            detail: TaskInput.fromObject({
              Message: 'Validation Exception',
            }),
            detailType: 'ValidationException',
            source: 'LocalTestingSource',
          },
        ],
      }
    );

    const validation = new sfn.Parallel(this, validationStateName, {
      resultSelector: {
        identityResult: JsonPath.stringAt('$[0].identity'),
        addressResult: JsonPath.stringAt('$[1].address'),
      },
      resultPath: JsonPath.stringAt('$.results'),
    });

    const detectSentimentState = new sfn.CustomState(
      this,
      detectSentimentStateName,
      {
        stateJson: {
          Type: 'Task',
          Resource: 'arn:aws:states:::aws-sdk:comprehend:detectSentiment',
          Parameters: {
            LanguageCode: 'en',
            'Text.$': JsonPath.stringAt('$.data.comments'),
          },
          ResultSelector: {
            'sentimentAnalysis.$': JsonPath.stringAt('$'),
          },
          ResultPath: JsonPath.stringAt('$.results'),
          Retry: [
            {
              ErrorEquals: ['InternalServerException'],
              IntervalSeconds: 1,
              MaxAttempts: 3,
              BackoffRate: 1,
            },
          ],
        },
      }
    );

    const negativeSentimentDetectedState = new tasks.EventBridgePutEvents(
      this,
      negativeSentimentDetectedStateName,
      {
        entries: [
          {
            detail: TaskInput.fromObject({
              Message: 'Negative Sentiment Detected',
              Data: JsonPath.stringAt('$.data'),
            }),
            detailType: 'NegativeSentiment',
            source: 'LocalTestingSource',
          },
        ],
      }
    );

    const addToFollowUpState = new tasks.DynamoPutItem(
      this,
      addToFollowUpStateName,
      {
        inputPath: '$.data',
        table: ddb.Table.fromTableName(this, 'FollowUpTable', 'FollowUpTable'),
        item: {
          PK: DynamoAttributeValue.fromString(
            JsonPath.stringAt('$.identity.email')
          ),
        },
        resultSelector: {
          dbUpdateStatusCode: JsonPath.stringAt(
            '$.SdkHttpMetadata.HttpStatusCode'
          ),
        },
        resultPath: JsonPath.stringAt('$.results'),
      }
    );

    const customerAddedToFollowupState = new tasks.EventBridgePutEvents(
      this,
      customerAddedToFollowupStateName,
      {
        inputPath: JsonPath.stringAt('$.data'),
        entries: [
          {
            detail: TaskInput.fromObject({
              Message: 'Customer Added for follow up',
              EmailAddress: JsonPath.stringAt('$.identity.email'),
            }),
            detailType: 'CustomerAdded',
            source: 'LocalTestingSource',
          },
        ],
      }
    );

    const isPositiveSentimentState = new sfn.Choice(
      this,
      isPositiveSentimentStateName
    );

    const definition = validation
      .branch(checkIdentityState, checkAddressState)
      .addCatch(customValidationFailedState, {
        errors: ['CustomValidationError'],
      })
      .addCatch(validationExceptionState, { errors: ['States.ALL'] })
      .next(detectSentimentState)
      .next(
        isPositiveSentimentState
          .when(
            sfn.Condition.stringEquals(
              JsonPath.stringAt('$.results.sentimentAnalysis.Sentiment'),
              'POSITIVE'
            ),
            addToFollowUpState.next(customerAddedToFollowupState)
          )
          .otherwise(negativeSentimentDetectedState)
      );

    const sm = new sfn.StateMachine(this, stateMachineName, {
      stateMachineName,
      definition,
    });
  }
}
