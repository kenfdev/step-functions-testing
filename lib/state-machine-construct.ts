import { Construct } from 'constructs';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { TaskInput } from 'aws-cdk-lib/aws-stepfunctions';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import { Duration } from 'aws-cdk-lib';

type StateMachineProps = {
  stateMachineName?: string;
  lambdaStateName?: string;
  sqsStateName?: string;
  retryInterval?: Duration;
};

export class StateMachineConstruct extends Construct {
  constructor(scope: Construct, id: string, props: StateMachineProps = {}) {
    super(scope, id);

    const {
      lambdaStateName = 'LambdaState',
      sqsStateName = 'SQSState',
      stateMachineName,
      retryInterval = Duration.seconds(2),
    } = props;

    const lambdaState = new tasks.LambdaInvoke(this, lambdaStateName, {
      lambdaFunction: lambda.Function.fromFunctionArn(
        this,
        'lambdaInvoke',
        'arn:aws:lambda:us-west-2:123456789012:function:my-function'
      ),
      payload: TaskInput.fromObject({
        'Payload.$': '$',
        FunctionName: 'HelloWorldFunction',
      }),
      retryOnServiceExceptions: false,
    }).addRetry({
      errors: ['States.ALL'],
      backoffRate: 2,
      maxAttempts: 3,
      interval: retryInterval,
    });

    const sqsState = new tasks.SqsSendMessage(this, sqsStateName, {
      queue: sqs.Queue.fromQueueArn(
        this,
        'sqs',
        'arn:aws:sqs:us-west-2:444455556666:queue1'
      ),
      messageBody: TaskInput.fromJsonPathAt('$'),
    });

    const definition = lambdaState.next(sqsState);

    new sfn.StateMachine(this, 'SampleStateMachine', {
      stateMachineName,
      definition,
    });
  }
}
