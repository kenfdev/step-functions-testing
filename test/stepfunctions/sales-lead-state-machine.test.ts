import {
  MockedResponse,
  StateMachineTestCase,
  StateMachineTestDefinition,
  StepFunctionsMockConfig,
} from '../helpers/mock';

import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import {
  createJestTestFromMockConfig,
  extractStateMachineAsls,
} from '../helpers/stepfunctions';
import { Duration } from 'aws-cdk-lib';
import { SalesLeadsStateMachineConstruct } from '../../lib/sales-leads-state-machine-construct';

// TODO: dummy code
type LambdaResponse = {
  StatusCode: number;
  Payload: {
    statusCode: number;
    body: any;
  };
};
type DynamoResponse = {
  SdkHttpMetadata: {
    HttpStatusCode: number;
  };
};
type EventBridgeResponse = {
  Payload: {
    Entries: { EventId: string }[];
    FailedEntryCount: number;
  };
};
type ComprehendResponse = {
  Sentiment: 'POSITIVE' | 'NEGATIVE';
  SentimentScore: {
    Mixed: number;
    Negative: number;
    Neutral: number;
    Positive: number;
  };
};

export const stateNames = {
  validationStateName: 'Validation',
  checkIdentityStateName: 'Check Identity',
  checkAddressStateName: 'Check Address',
  customValidationFailedStateName: 'CustomValidationFailed',
  validationExceptionStateName: 'ValidationException',
  detectSentimentStateName: 'DetectSentiment',
  isPositiveSentimentStateName: 'Is Positive Sentiment?',
  negativeSentimentDetectedStateName: 'NegativeSentimentDetected',
  addToFollowUpStateName: 'Add to FollowUp',
  customerAddedToFollowupStateName: 'CustomerAddedToFollowup',
} as const;

type StateNameKeys = keyof typeof stateNames;
type StateName = typeof stateNames[StateNameKeys];

const stateMachineName = 'LocalTesting';

const checkIdentityLambdaMockedSuccess = new MockedResponse(
  'CheckIdentityLambdaMockedSuccess'
).return<LambdaResponse>({
  StatusCode: 200,
  Payload: {
    statusCode: 200,
    body: JSON.stringify({
      approved: true,
      message: 'identity validation passed',
    }),
  },
});

const checkAddressLambdaMockedSuccess = new MockedResponse(
  'CheckAddressLambdaMockedSuccess'
).return<LambdaResponse>({
  StatusCode: 200,
  Payload: {
    statusCode: 200,
    body: JSON.stringify({
      approved: true,
      message: 'address validation passed',
    }),
  },
});

const addToFollowUpSuccess = new MockedResponse(
  'AddToFollowUpSuccess'
).return<DynamoResponse>({
  SdkHttpMetadata: {
    HttpStatusCode: 200,
  },
});

const customerAddedToFollowupSuccess = new MockedResponse(
  'CustomerAddedToFollowupSuccess'
).return({
  StatusCode: 200,
  Payload: {
    statusCode: 200,
  },
});

const checkIdentityLambdaMockedThrowError = new MockedResponse(
  'CheckIdentityLambdaMockedThrowError'
).throw('CustomValidationError', 'Check Identity Validation Failed', 3);

const checkAddressLambdaMockedThrowExceptionSuccess = new MockedResponse(
  'CheckAddressLambdaMockedThrowExceptionSuccess'
).throw('CustomAddressValidationError', 'Address Validation Exception');

const customValidationFailedPutEventSuccess = new MockedResponse(
  'CustomValidationFailedPutEventSuccess'
).return<EventBridgeResponse>({
  Payload: {
    Entries: [
      {
        EventId: 'abc123',
      },
    ],
    FailedEntryCount: 0,
  },
});

const validationExceptionPutEventSuccess = new MockedResponse(
  'ValidationExceptionPutEventSuccess'
).return<EventBridgeResponse>({
  Payload: {
    Entries: [
      {
        EventId: 'abc123',
      },
    ],
    FailedEntryCount: 0,
  },
});

const detectSentimentPositive = new MockedResponse(
  'DetectSentimentPositive'
).return<ComprehendResponse>({
  Sentiment: 'POSITIVE',
  SentimentScore: {
    Mixed: 0.00012647535,
    Negative: 0.00008031699,
    Neutral: 0.0051454515,
    Positive: 0.9946478,
  },
});

const detectSentimentNegative = new MockedResponse(
  'DetectSentimentNegative'
).return<ComprehendResponse>({
  Sentiment: 'NEGATIVE',
  SentimentScore: {
    Mixed: 0.00012647535,
    Positive: 0.00008031699,
    Neutral: 0.0051454515,
    Negative: 0.9946478,
  },
});

const negativeSentimentDetectedSuccess = new MockedResponse(
  'NegativeSentimentDetectedSuccess'
).return<EventBridgeResponse>({
  Payload: {
    Entries: [
      {
        EventId: 'abc123',
      },
    ],
    FailedEntryCount: 0,
  },
});

const detectSentimentRetryOnErrorWithSuccess = new MockedResponse(
  'DetectSentimentRetryOnErrorWithSuccess'
)
  .throw(
    'InternalServerException',
    'Server Exception while calling DetectSentiment API in Comprehend Service',
    2
  )
  .return<ComprehendResponse>({
    Sentiment: 'POSITIVE',
    SentimentScore: {
      Mixed: 0.00012647535,
      Negative: 0.00008031699,
      Neutral: 0.0051454515,
      Positive: 0.9946478,
    },
  });

const input = {
  data: {
    firstname: 'Jane',
    lastname: 'Doe',
    identity: {
      email: 'jdoe@example.com',
      ssn: '123-45-6789',
    },
    address: {
      street: '123 Main St',
      city: 'Columbus',
      state: 'OH',
      zip: '43219',
    },
    comments:
      'I am glad to sign-up for this service. Looking forward to different options.',
  },
};

const stateMachineTestDefinition = new StateMachineTestDefinition(
  stateMachineName
)
  .addTestCase(
    new StateMachineTestCase<StateName>('HappyPathTest')
      .withInput(input)
      .addMockedState(
        stateNames.checkIdentityStateName,
        checkIdentityLambdaMockedSuccess
      )
      .addMockedState(
        stateNames.checkAddressStateName,
        checkAddressLambdaMockedSuccess
      )
      .addMockedState(
        stateNames.detectSentimentStateName,
        detectSentimentPositive
      )
      .addMockedState(stateNames.addToFollowUpStateName, addToFollowUpSuccess)
      .addMockedState(
        stateNames.customerAddedToFollowupStateName,
        customerAddedToFollowupSuccess
      )
  )
  .addTestCase(
    new StateMachineTestCase<StateName>('NegativeSentimentTest')
      .withInput(input)
      .addMockedState(
        stateNames.checkIdentityStateName,
        checkIdentityLambdaMockedSuccess
      )
      .addMockedState(
        stateNames.checkAddressStateName,
        checkAddressLambdaMockedSuccess
      )
      .addMockedState(
        stateNames.detectSentimentStateName,
        detectSentimentNegative
      )
      .addMockedState(
        stateNames.negativeSentimentDetectedStateName,
        negativeSentimentDetectedSuccess
      )
  );

const config = new StepFunctionsMockConfig();
config.addTestDefinition(stateMachineTestDefinition);

const stack = new cdk.Stack();
new SalesLeadsStateMachineConstruct(stack, 'SalesLeadsStateMachine', {
  validationStateName: 'Validation',
  checkIdentityStateName: 'Check Identity',
  checkAddressStateName: 'Check Address',
  customValidationFailedStateName: 'CustomValidationFailed',
  validationExceptionStateName: 'ValidationException',
  detectSentimentStateName: 'DetectSentiment',
  isPositiveSentimentStateName: 'Is Positive Sentiment?',
  negativeSentimentDetectedStateName: 'NegativeSentimentDetected',
  addToFollowUpStateName: 'Add to FollowUp',
  customerAddedToFollowupStateName: 'CustomerAddedToFollowup',
  stateMachineName: stateMachineName,
  retryInterval: Duration.seconds(0),
});
const cfnStackJson = Template.fromStack(stack).toJSON();
// extract the Asl part
const asls = extractStateMachineAsls(cfnStackJson);

describe(
  'SalesLeasStateMachine',
  createJestTestFromMockConfig(config, asls[0])
);
