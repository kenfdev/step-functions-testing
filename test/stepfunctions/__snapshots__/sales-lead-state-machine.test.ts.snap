// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalesLeasStateMachine LocalTesting HappyPathTest 1`] = `
Map {
  "Validation" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"}}",
    },
    Object {
      "output": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"addressResult\\":{\\"approved\\":true,\\"message\\":\\"address validation passed\\"},\\"identityResult\\":{\\"approved\\":true,\\"message\\":\\"identity validation passed\\"}}}",
    },
  ],
  "Check Identity" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"}}",
    },
    Object {
      "output": "{\\"identity\\":{\\"approved\\":true,\\"message\\":\\"identity validation passed\\"}}",
    },
  ],
  "Check Address" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"}}",
    },
    Object {
      "output": "{\\"address\\":{\\"approved\\":true,\\"message\\":\\"address validation passed\\"}}",
    },
  ],
  "DetectSentiment" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"addressResult\\":{\\"approved\\":true,\\"message\\":\\"address validation passed\\"},\\"identityResult\\":{\\"approved\\":true,\\"message\\":\\"identity validation passed\\"}}}",
    },
    Object {
      "output": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"sentimentAnalysis\\":{\\"Sentiment\\":\\"POSITIVE\\",\\"SentimentScore\\":{\\"Mixed\\":1.2647535E-4,\\"Negative\\":8.031699E-5,\\"Neutral\\":0.0051454515,\\"Positive\\":0.9946478}}}}",
    },
  ],
  "Is Positive Sentiment?" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"sentimentAnalysis\\":{\\"Sentiment\\":\\"POSITIVE\\",\\"SentimentScore\\":{\\"Mixed\\":1.2647535E-4,\\"Negative\\":8.031699E-5,\\"Neutral\\":0.0051454515,\\"Positive\\":0.9946478}}}}",
    },
    Object {
      "output": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"sentimentAnalysis\\":{\\"Sentiment\\":\\"POSITIVE\\",\\"SentimentScore\\":{\\"Mixed\\":1.2647535E-4,\\"Negative\\":8.031699E-5,\\"Neutral\\":0.0051454515,\\"Positive\\":0.9946478}}}}",
    },
  ],
  "Add to FollowUp" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"sentimentAnalysis\\":{\\"Sentiment\\":\\"POSITIVE\\",\\"SentimentScore\\":{\\"Mixed\\":1.2647535E-4,\\"Negative\\":8.031699E-5,\\"Neutral\\":0.0051454515,\\"Positive\\":0.9946478}}}}",
    },
    Object {
      "output": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"dbUpdateStatusCode\\":200}}",
    },
  ],
  "CustomerAddedToFollowup" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"dbUpdateStatusCode\\":200}}",
    },
    Object {
      "output": "{\\"StatusCode\\":200,\\"Payload\\":{\\"statusCode\\":200}}",
    },
  ],
}
`;

exports[`SalesLeasStateMachine LocalTesting NegativeSentimentTest 1`] = `
Map {
  "Validation" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"}}",
    },
    Object {
      "output": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"addressResult\\":{\\"approved\\":true,\\"message\\":\\"address validation passed\\"},\\"identityResult\\":{\\"approved\\":true,\\"message\\":\\"identity validation passed\\"}}}",
    },
  ],
  "Check Identity" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"}}",
    },
    Object {
      "output": "{\\"identity\\":{\\"approved\\":true,\\"message\\":\\"identity validation passed\\"}}",
    },
  ],
  "Check Address" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"}}",
    },
    Object {
      "output": "{\\"address\\":{\\"approved\\":true,\\"message\\":\\"address validation passed\\"}}",
    },
  ],
  "DetectSentiment" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"addressResult\\":{\\"approved\\":true,\\"message\\":\\"address validation passed\\"},\\"identityResult\\":{\\"approved\\":true,\\"message\\":\\"identity validation passed\\"}}}",
    },
    Object {
      "output": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"sentimentAnalysis\\":{\\"Sentiment\\":\\"NEGATIVE\\",\\"SentimentScore\\":{\\"Mixed\\":1.2647535E-4,\\"Positive\\":8.031699E-5,\\"Neutral\\":0.0051454515,\\"Negative\\":0.9946478}}}}",
    },
  ],
  "Is Positive Sentiment?" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"sentimentAnalysis\\":{\\"Sentiment\\":\\"NEGATIVE\\",\\"SentimentScore\\":{\\"Mixed\\":1.2647535E-4,\\"Positive\\":8.031699E-5,\\"Neutral\\":0.0051454515,\\"Negative\\":0.9946478}}}}",
    },
    Object {
      "output": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"sentimentAnalysis\\":{\\"Sentiment\\":\\"NEGATIVE\\",\\"SentimentScore\\":{\\"Mixed\\":1.2647535E-4,\\"Positive\\":8.031699E-5,\\"Neutral\\":0.0051454515,\\"Negative\\":0.9946478}}}}",
    },
  ],
  "NegativeSentimentDetected" => Array [
    Object {
      "input": "{\\"data\\":{\\"firstname\\":\\"Jane\\",\\"lastname\\":\\"Doe\\",\\"identity\\":{\\"email\\":\\"jdoe@example.com\\",\\"ssn\\":\\"123-45-6789\\"},\\"address\\":{\\"street\\":\\"123 Main St\\",\\"city\\":\\"Columbus\\",\\"state\\":\\"OH\\",\\"zip\\":\\"43219\\"},\\"comments\\":\\"I am glad to sign-up for this service. Looking forward to different options.\\"},\\"results\\":{\\"sentimentAnalysis\\":{\\"Sentiment\\":\\"NEGATIVE\\",\\"SentimentScore\\":{\\"Mixed\\":1.2647535E-4,\\"Positive\\":8.031699E-5,\\"Neutral\\":0.0051454515,\\"Negative\\":0.9946478}}}}",
    },
    Object {
      "output": "{\\"Payload\\":{\\"Entries\\":[{\\"EventId\\":\\"abc123\\"}],\\"FailedEntryCount\\":0}}",
    },
  ],
}
`;
